x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/suggestion_panel.tsx

```diff
@@ -117,70 +138,241 @@ function InnerSuggestionPanel({
   dispatch,
   frame,
   ExpressionRenderer: ExpressionRendererComponent,
+  stagedPreview,
 }: SuggestionPanelProps) {
-  if (!activeDatasourceId) {
-    return null;
-  }
+  const currentDatasourceStates = stagedPreview ? stagedPreview.datasourceStates : datasourceStates;
+  const currentVisualizationState = stagedPreview
+    ? stagedPreview.visualization.state
+    : visualizationState;
+  const currentVisualizationId = stagedPreview
+    ? stagedPreview.visualization.activeId
+    : activeVisualizationId;
+
+  const { suggestions, currentStateExpression } = useMemo(() => {
+    const newSuggestions = getSuggestions({
+      datasourceMap,
+      datasourceStates: currentDatasourceStates,
+      visualizationMap,
+      activeVisualizationId: currentVisualizationId,
+      visualizationState: currentVisualizationState,
+    })
+      .map(suggestion => ({
+        ...suggestion,
+        previewExpression: preparePreviewExpression(
+          suggestion,
+          visualizationMap[suggestion.visualizationId],
+          datasourceMap,
+          currentDatasourceStates,
+          frame
+        ),
+      }))
+      .filter(suggestion => !suggestion.hide)
+      .slice(0, MAX_SUGGESTIONS_DISPLAYED);
 
-  const suggestions = getSuggestions({
+    const newStateExpression =
+      currentVisualizationState && currentVisualizationId
+        ? preparePreviewExpression(
+            { visualizationState: currentVisualizationState },
+            visualizationMap[currentVisualizationId],
+            datasourceMap,
+            currentDatasourceStates,
+            frame
+          )
+        : undefined;
+
+    return { suggestions: newSuggestions, currentStateExpression: newStateExpression };
+  }, [
+    currentDatasourceStates,
+    currentVisualizationState,
+    currentVisualizationId,
     datasourceMap,
-    datasourceStates,
     visualizationMap,
-    activeVisualizationId,
-    visualizationState,
-  })
-    .filter(suggestion => !suggestion.hide)
-    .slice(0, MAX_SUGGESTIONS_DISPLAYED);
+  ]);
+
+  const [lastSelectedSuggestion, setLastSelectedSuggestion] = useState<number>(-1);
+
+  useEffect(() => {
+    // if the staged preview is overwritten by a suggestion,
+    // reset the selected index to "current visualization" because
+    // we are not in transient suggestion state anymore
+    if (!stagedPreview && lastSelectedSuggestion !== -1) {
+      setLastSelectedSuggestion(-1);
+    }
+  }, [stagedPreview]);
+
+  if (!activeDatasourceId) {
+    return null;
+  }
 
   if (suggestions.length === 0) {
     return null;
   }
 
+  function rollbackToCurrentVisualization() {
+    if (lastSelectedSuggestion !== -1) {
+      setLastSelectedSuggestion(-1);
+      dispatch({
+        type: 'ROLLBACK_SUGGESTION',
+      });
+    }
+  }
+
+  const expressionContext = {
+    query: frame.query,
+    timeRange: {
+      from: frame.dateRange.fromDate,
```

Comment:
Unless there's a good reason not to, I think we should rename `from` to `fromDate` and similar for `to`, just to be consistent in our naming everywhere.

Reply:
Agreed, but we should do it the other way around, because `from` and `to` is the "official" way (inherited from `TimeRange` straight out of `'src/plugins/data/public';`. Out of scope for this PR, but I added it to the technical debt list.